[Schemas] Implement a connector utility service (no-op implementation)

hariso opened the pull request:
### Description

Part of https://github.com/ConduitIO/conduit/issues/1561.

### Quick checks:

- [ ] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [ ] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-protocol/pulls) for the same update/change.
- [ ] I have written unit tests.
- [ ] I have made sure that the PR is of reasonable size and can be easily reviewed.

lovromazgon commented:
@hariso FYI, this is relevant: https://github.com/hashicorp/go-plugin/tree/main/examples/bidirectional

Makes me think that we should move this file into `connector/v2` (which would mean you would need to wait for my PR).

hariso commented:
> @hariso FYI, this is relevant: https://github.com/hashicorp/go-plugin/tree/main/examples/bidirectional
> 
> Makes me think that we should move this file into `connector/v2` (which would mean you would need to wait for my PR).

Sweet, looks interesting! I'll look into if/how we can possibly use it!

lovromazgon commented on proto/schema/v1/schema.proto:
```diff
@@ -0,0 +1,41 @@
+package schema.v1;
+
+service SchemaService {
+  rpc Register(RegisterRequest) returns (RegisterResponse);
+  rpc Fetch(FetchRequest) returns (FetchResponse);
```
I propose sticking to the same terminology we use for example in the [Conduit API](https://github.com/ConduitIO/conduit/blob/main/proto/api/v1/api.proto), i.e. `Create`, `Get`, `Update`, `Delete`.

hariso replied:
Sounds good!

lovromazgon commented on proto/schema/v1/schema.proto:
```diff
@@ -0,0 +1,41 @@
+
+  // The schema's globally unique ID
+  string id = 1;
+  // The schema's name (globally unique)
+  string name = 2;
+  int32 version = 3;
```
The name can probably be repeated in other schemas, but the name+version combination should be unique.

lovromazgon commented on proto/schema/v1/schema.proto:
```diff
@@ -0,0 +1,41 @@
+  bytes bytes = 3;
+}
+message RegisterResponse {
+  string id = 1;
```
How about we return the whole created schema back? When working with schema registries in the past I found it frustrating that I had to create a schema and then right after fetch the schema if I wanted to know for example the version that was assigned to the schema.

hariso replied:
I was thinking about that too, but wasn't sure if there was a use case... In a source connector's context, when would it want to know the created schema's version? On the other hand, it doesn't hurt having it, and we don't loose anything.:)

lovromazgon replied:
If we add it, we'll never need it. If we don't, we'll need it as soon as we cut a release and changing this would create a breaking change 😄

hariso replied:
Exactly!

lovromazgon commented on proto/schema/v1/schema.proto:
```diff
@@ -0,0 +1,41 @@
+syntax = "proto3";
+
+package schema.v1;
```
I'm not sure if this warrants a whole independent package. I'd argue this is part of the interface between Conduit and the connector, so we could put it into `connector.v2`. Let me explain.

Let's think about the backwards compatibility that I added in https://github.com/ConduitIO/conduit-connector-protocol/pull/90. Updated connectors that use protocol v2 will work with older versions of Conduit, as they can fall back to protocol v1. The problem is that new connectors might also require this schema service, so they won't work correctly with older versions of Conduit. So maybe we should remove the backwards compatibility? 🤔 And this same reason is what makes me think that this belongs in `connector.v2`, as it makes it clear that the schema service is only available to connectors that use the connector protocol v2.

As an alternative, we could also put it into a "connectors utilities" package, where we can add more utilities in the future (like storage). This would mean that if we evolve the utilities and introduce a breaking change there, we don't have to necessarily bump the connector proto.

raulb replied:
@lovromazgon my first question is why to tie it to the connectors package at all? Maybe these could still be separate packages and say that connector.v2 will be compatible with schema.v1 being that the main reason to maybe upgrade to connector.v2 if they're still on connector.v1.

What do you think?

hariso replied:
A question first: what is `connector.v2` supposed to tell us? Is it v2 of the connector API, or is the connector part of protocol v2? I was assuming it's the first one (version 2 of connector API), because if we were versioning the whole protocol, I'd expect a structure like this:
```
# v1 of the protocol supports connectors and foos
v1
  - connector
  - foo
# in v2 of the protocol we introduced bar
v2:
  - connector
  - foo
  - bar
```

Also, my understanding of the `connector` folder is that it contains definitions of connector services, not necessarily everything a connector might use.

Based on the above assumptions I made the decision to create a schema.v1 package. I wouldn't like the schema service to be "too close" to the connector definitions, because they can evolve for different reasons and at different paces. 

Generally, I think putting it into a "connector utilities" is a good thing too. The only thing I don't like about it is that it gives the impression that schemas are for connectors only. Schemas as a feature will be used in processors, and if we had gRPC processors those would use it too.

lovromazgon replied:
> Generally, I think putting it into a "connector utilities" is a good thing too. The only thing I don't like about it is that it gives the impression that schemas are for connectors only. Schemas as a feature will be used in processors, and if we had gRPC processors those would use it too.

Let's address this part first, so we're on the same page. We are in the `conduit-connector-protocol` repo, with emphasis on `connector` 🙃 As the readme says:

  _This repository is the only connection point between **Conduit** and a **connector**._

Stuff in this repo should _only_ be relevant to connectors, not processors, not the public API. If/when we expose schemas to processors, I'll argue that we should do it through WASM host functions, since that's the mechanism we use to communicate between a processor and Conduit. It would be a bit strange if a processor, running in the same process as Conduit, would communicate with Conduit through gRPC. As for the public API - that one has a different purpose and might need to be compatible with the Confluent API, if/when we decide to expose it to end-users.

> what is `connector.v2` supposed to tell us?

I see it as the definition of the interface between Conduit and the connector. And by interface I was thinking both ways - what Conduit exposes, and what the connector exposes.

But maybe we could say that `connector.v2` defines the services that the connector implements and we put this schema service into the package `conduit.v1`, since it's a service that Conduit exposes. So connector -> `connector.v2`, Conduit -> `conduit.v1`, both can evolve separately.

As for the versioning, we've been following the `FILE` protobuf [breaking rules](https://buf.build/docs/breaking/rules) (we have a CI action that checks this). It's completely valid and _not_ a breaking change if a new `rpc` gets added to the service. For instance, we did this with lifecycle events in the past. Conduit executes the lifecycle functions _if they exist_, otherwise it ignores them, as it's apparently an older connector. The problem appears if the opposite happens, e.g. we add a new `rpc` to the Conduit utilities service. That change is backwards compatible to old consumers (e.g. old plugins will still work fine, they don't care about the added `rpc`), but new consumers that start using the added `rpc` _will not_ be backwards compatible anymore, since they won't find the `rpc` in an older version of the service.

Why do we care? Because we don't want to find out a connector doesn't work when we start a pipeline and it suddenly breaks when it tries to hit an `rpc` that doesn't exist. Ideally, we should catch this early, e.g. when an older version of Conduit tries to register a connector, that expects the new version of a utility service, it should already fail to be registered.

I think we won't get around some kind of "handshake" between the connector and Conduit when the plugin gets registered, where the plugin can reject being registered if Conduit doesn't expose the expected services/rpcs. As mentioned in my previous comment, we could bump the major version of the whole proto package (e.g. `conduit.v1` -> `conduit.v2`), but that sounds wrong and overkill for changes that are essentially backwards compatible from the perspective of old consumers.

hariso replied:
Thanks for the detailed comment, Lovro! It cleared up some things. 

I was always reading the repo and its purpose as "a Conduit connector's protocol". And the protocol (services) defined in this repo are the only way to communicate with a connector. I was always reasoning about Conduit and a connector as being like very distinct (micro-)services, each with its own API. All (and only) the APIs Conduit exposes are defined in the Conduit repo, all (and only) connector APIs are in this repo. Anyway...

With things being as mentioned above, having something like `conduit.v1` makes sense to me.

Also, for the record: on the standup today, we discussed failing as early as possible when an older version of Conduit tries to load a newer version of a connector. While it's not a must-have for the schema support epic, it's still a very good improvement of user experience, and we'd like to have it. For now, we're leaning towards a solution that will be discussed in more details separately.

hariso replied:
Also, for the record: on the standup today, we discussed failing as early as possible when an older version of Conduit tries to load a newer version of a connector. While it's not a must-have for the schema support epic, it's still a very good improvement of user experience, and we'd like to have it. For now, we're leaning towards a solution that will be discussed in more details separately.

lovromazgon replied:
I've been thinking more about this and I think this could be addressed by expanding the specification. The specification would include a list of services and their versions, so Conduit itself could check if it provides everything the connector expects when it fetches the specifications (this happens when registering the plugin). An additional benefit is that if we extract the [connector specifications into a YAML file](https://github.com/ConduitIO/conduit/discussions/1522), we can know if the connector is compatible with a specific Conduit version only by looking into the specification and we can display a compatibility matrix (or minimum Conduit version) in the docs.

So in the end, I think this can be safely added afterwards, all we need is a way for [specgen](https://github.com/ConduitIO/conduit/discussions/1522), to figure out what utilities the connector is using, then it can include those dependencies in the specifications yaml, everything else can happen in Conduit. This can still be done using the approach we discussed, by putting the utilities for a specific service into a separate package in the SDK.

hariso replied:
That sounds like a good idea! :+1:  It's similar to what we were discussing earlier about features.

lovromazgon commented on proto/schema/v1/schema.proto:
```diff
@@ -0,0 +1,41 @@
+
+package schema.v1;
+
+service SchemaService {
```
I'm also thinking that it would be good if we internally version the schema service (or any connector utilities we add in the future) and allow the connector to fetch the version.

Imagine we add a function to delete a schema in the next release. Connectors that start using that function are not backwards compatible with older versions of Conduit anymore. So maybe those connectors could specify that they require a minimum version of the `SchemaService` to work correctly, and the connector refuses to be even loaded into Conduit if it doesn't provide the minimum version.

We could do it using the go-plugin functionality for versioning plugins (basically what we did now when we introduced v2), but then we would have to bump the connector protocol every time we add a function + add all those proto transforms. We could instead do it manually by having a `rpc Version();` on the schema service (and any utility service we add in the future). This would give us more fine-grained control over what is required by the connector, but we would also need to manually write the code to figure out compatibility.

LMK what you think about this.

raulb replied:
> We could do it using the go-plugin functionality for versioning plugins (basically what we did now when we introduced v2), but then we would have to bump the connector protocol every time we add a function + add all those proto transforms.

My vote goes to this option.

lovromazgon replied:
I expanded on this a bit in [another comment](https://github.com/ConduitIO/conduit-connector-protocol/pull/88#discussion_r1611777494).

lovromazgon commented on conduit/v1/fromproto/schema.go:
```diff
@@ -0,0 +1,44 @@
+	conduitv1 "github.com/conduitio/conduit-connector-protocol/proto/conduit/v1"
+)
+
+func SchemaInstance(req *conduitv1.CreateSchemaRequest) (schema.Instance, error) {
```
I propose to introduce types for `CreateSchemaRequest` (and other requests and responses) in `github.com/conduitio/conduit-connector-protocol/conduit/v1`, to follow the same pattern we use in `cplugin` (for example [here](https://github.com/ConduitIO/conduit-connector-protocol/blob/4230db1b457029e74c11f252e4019a5eccd25543/cplugin/source.go#L58-L61)). Even if the type is empty or just contains a single field, it makes the code explicit. If we set up a strict structure now I believe it will be easier to work with this down the road. Currently for example it's not exactly clear that we only got 3 fields in the request, since `schema.Instance` contains more fields than that.

That said, in the create response we can reference a `schema.Instance` in the struct, since the proto message reflects exactly the same type (and that shouldn't change).

Btw, we'll also need proto conversions for responses, we might as well add them now already. We'll need them in the SDK.

hariso replied:
Done!

lovromazgon commented on conduit/v1/fromproto/schema.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+func SchemaType(typ conduitv1.Schema_Type) (schema.Type, error) {
+	switch typ {
+	case conduitv1.Schema_TYPE_AVRO:
+		return schema.TypeAvro, nil
+	default:
+		return 0, fmt.Errorf("unsupported %q", typ)
+	}
```
There's a trick using compile time checks that we can use to ensure the proto enum matches the Go enum, then the conversion can be as simple as `schema.Type(typ)` (see the [checks for operations](https://github.com/ConduitIO/conduit-commons/blob/bf00c71a3ff89864a24c54fca26a3601243d1c50/opencdc/proto.go#L24-L31) and how we [transform](https://github.com/ConduitIO/conduit-commons/blob/bf00c71a3ff89864a24c54fca26a3601243d1c50/opencdc/proto.go#L61) it).

hariso replied:
Ah, good callout, I forgot about that one!

lovromazgon commented on conduit/v1/toproto/schema.go:
```diff
@@ -0,0 +1,55 @@
+	conduitv1 "github.com/conduitio/conduit-connector-protocol/proto/conduit/v1"
+)
+
+func GetResponse(inst schema.Instance) *conduitv1.GetSchemaResponse {
```
One request around naming - the other `fromproto` and `toproto` packages in this repo follow a strict naming scheme, where the name of the function is the same as the type name it's converting. So if we introduce `conduit.GetSchemaResponse` as I propose in the other comment, then this would be changed to:

```suggestion
func GetSchemaResponse(in conduit.GetSchemaResponse) *conduitv1.GetSchemaResponse {
```

lovromazgon reviewed (commented):
I hope I'm not inhibiting your creative freedom, but you know how much I like consistency 😄 So my comments come from the perspective of having the same approach in this package as in `cplugin`.

raulb commented on go.mod:
```diff
@@ -244,12 +244,14 @@ require (
 	mvdan.cc/unparam v0.0.0-20240427195214-063aff900ca1 // indirect
 )
+
+replace github.com/conduitio/conduit-commons => ../conduit-commons
```
We should remove this one before merging.

hariso replied:
Good callout!

raulb commented on conduit/schema/v1/grpc_client.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+func NewGRPCSchemaClient() (*GRPCSchemaClient, error) {
+	conn, err := grpc.NewClient("localhost:8085")
```
Looking today more in depth the connector-protocol code I wondered if we should these ports configurable in case of collision.

hariso replied:
You're right about that. I've added that as part of https://github.com/ConduitIO/conduit/issues/1563.

hariso commented:
@lovromazgon @raulb Thanks for all the feedback so far! Please take another look, I think I addressed all the comments. I've also changed the target branch to a temporary one, so it's easier to review. Once all of the recent refactoring is merged, I'll change the base to `main` again.

raulb reviewed (commented):
@hariso looks good! Only left a couple of questions.

hariso commented on proto/conduit/v1/schema.proto:
```diff
@@ -0,0 +1,42 @@
+
+message Schema {
+  enum Type {
+    TYPE_UNSPECIFIED = 0;
```
It's about a best practice that is described [here](https://protobuf.dev/programming-guides/style/#enums) and [here](https://protobuf.dev/programming-guides/dos-donts/#unspecified-enum).

hariso commented on conduit/schema/v1/toproto/schema.go:
```diff
@@ -0,0 +1,59 @@
+func CreateSchemaRequest(request v1.CreateRequest) *conduitv1.CreateSchemaRequest {
+	return &conduitv1.CreateSchemaRequest{
+		Name:  request.Name,
+		Type:  conduitv1.Schema_TYPE_AVRO,
```
The intention was to make sure all Avro schemas go in... But now that you bring this up, it's probably better to have a type field, but only allow one value, i.e. Avro. I'll change, thanks for calling this out!
